### PR TITLE
[SPARK-22529] [SQL] Relation stats should be consistent with other plans based on cbo config

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

Currently, relation stats is the same whether cbo is enabled or not. While relation (`LogicalRelation` or `HiveTableRelation`) is a `LogicalPlan`, its behavior is inconsistent with other plans. This can cause confusion when user runs EXPLAIN COST commands. Besides, when CBO is disabled, we apply the size-only estimation strategy, so there's no need to propagate other catalog statistics to relation.

## How was this patch tested?

Enhanced existing tests case and added a test case.

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @cloud-fan @gatorsmile

**Comment by wzhfy** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala [hunk 1]
```diff
@@ -95,6 +95,10 @@ case class SetCommand(kv: Option[(String, Option[String])]) extends RunnableComm
             s"prefix spark.hadoop (e.g., spark.hadoop.$key) when starting a Spark application. " +
             "For details, see the link: https://spark.apache.org/docs/latest/configuration.html#" +
             "dynamically-loading-spark-properties.")
+        } else if (key == SQLConf.CBO_ENABLED.key &&
+            value != sparkSession.conf.get(SQLConf.CBO_ENABLED.key)) {
+          // To change CBO config, first we need to invalidate stats for all relation caches.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala]: what about the stats in hive metastore?

**Comment by wzhfy** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -366,10 +366,16 @@ case class CatalogStatistics(
    * Convert [[CatalogStatistics]] to [[Statistics]], and match column stats to attributes based
    * on column names.
    */
-  def toPlanStats(planOutput: Seq[Attribute]): Statistics = {
-    val matched = planOutput.flatMap(a => colStats.get(a.name).map(a -> _))
-    Statistics(sizeInBytes = sizeInBytes, rowCount = rowCount,
-      attributeStats = AttributeMap(matched))
+  def toPlanStats(planOutput: Seq[Attribute], cboEnabled: Boolean): Statistics = {
+    val attrStats = planOutput.flatMap(a => colStats.get(a.name).map(a -> _))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: move into `if`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala [hunk 3]
```diff
@@ -29,21 +30,30 @@ class HiveExplainSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
...
-    checkKeywordsExist(sql("EXPLAIN COST  SELECT * FROM src "), "sizeInBytes", "rowCount")
+      // Has both sizeInBytes and rowCount after ANALYZE command
+      sql("ANALYZE TABLE src COMPUTE STATISTICS")
+      checkKeywordsExist(sql(explainCostCommand), "sizeInBytes", "rowCount")
+    }
+
+    withSQLConf(SQLConf.CBO_ENABLED.key -> "false") {
+      // Don't show rowCount if cbo is disabled
+      checkKeywordsExist(sql(explainCostCommand), "sizeInBytes")
+      checkKeywordsNotExist(sql(explainCostCommand), "rowCount")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala]: did you assume there is no table relation cache in this test?

Comment (wzhfy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala]: yes, should I refresh it for robustness?

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveExplainSuite.scala]: yes please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -366,10 +366,16 @@ case class CatalogStatistics(
...
-      attributeStats = AttributeMap(matched))
+  def toPlanStats(planOutput: Seq[Attribute], cboEnabled: Boolean): Statistics = {
+    if (cboEnabled) {
+      val attrStats = planOutput.flatMap(a => colStats.get(a.name).map(a -> _))
+      Statistics(sizeInBytes = sizeInBytes, rowCount = rowCount,
+        attributeStats = AttributeMap(attrStats))
+    } else {
+      // When CBO is disabled, we apply the size-only estimation strategy, so there's no need to
+      // propagate other statistics from catalog to the plan.
+      Statistics(sizeInBytes = sizeInBytes)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: If `rowCount` is available, why we ignore them?

**Pull request closed without merging**